File: src/com/maddyhome/idea/vim/action/copy/YankVisualAction.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2019 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.maddyhome.idea.vim.action.copy;

import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.editor.Caret;
import com.intellij.openapi.editor.Editor;
import com.maddyhome.idea.vim.action.VimCommandAction;
import com.maddyhome.idea.vim.command.Command;
import com.maddyhome.idea.vim.command.CommandFlags;
import com.maddyhome.idea.vim.command.MappingMode;
import com.maddyhome.idea.vim.common.TextRange;
import com.maddyhome.idea.vim.group.copy.YankCopyGroup;
import com.maddyhome.idea.vim.handler.VisualOperatorActionBatchHandler;
import com.maddyhome.idea.vim.helper.VimSelection;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.EnumSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * @author vlan
 */
public class YankVisualAction extends VimCommandAction {
  public YankVisualAction() {
    super(new VisualOperatorActionBatchHandler() {
      @Override
      public boolean executeForAllCarets(@NotNull Editor editor,
                                         @NotNull DataContext context,
                                         @NotNull Command cmd, @NotNull Map<Caret, VimSelection> caretsAndSelections) {
        Collection<? extends VimSelection> selections = caretsAndSelections.values();

        List<Integer> starts = new ArrayList<>();
        List<Integer> ends = new ArrayList<>();
        selections.forEach(selection -> {
          TextRange textRange = selection.toVimTextRange();
          Arrays.stream(textRange.getStartOffsets()).boxed().forEach(starts::add);
          Arrays.stream(textRange.getEndOffsets()).boxed().forEach(ends::add);
        });
        VimSelection vimSelection = selections.stream().findFirst().orElse(null);
        if (vimSelection == null) return false;
        int[] startsArray = starts.stream().mapToInt(i -> i).toArray();
        int[] endsArray = ends.stream().mapToInt(i -> i).toArray();
        return YankCopyGroup.INSTANCE
          .yankRange(editor, new TextRange(startsArray, endsArray), vimSelection.getType(), true);
      }
    });
  }

  @NotNull
  @Override
  public Set<MappingMode> getMappingModes() {
    return MappingMode.V;
  }

  @NotNull
  @Override
  public Set<List<KeyStroke>> getKeyStrokesSet() {
    return parseKeysSet("y");
  }

  @NotNull
  @Override
  public Command.Type getType() {
    return Command.Type.COPY;
  }

  @Override
  public EnumSet<CommandFlags> getFlags() {
    return EnumSet.of(CommandFlags.FLAG_EXIT_VISUAL);
  }
}

File: vim-engine/src/main/kotlin/com/maddyhome/idea/vim/action/copy/YankVisualAction.kt
/*
 * Copyright 2003-2023 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package com.maddyhome.idea.vim.action.copy

import com.intellij.vim.annotations.CommandOrMotion
import com.intellij.vim.annotations.Mode
import com.maddyhome.idea.vim.api.ExecutionContext
import com.maddyhome.idea.vim.api.VimCaret
import com.maddyhome.idea.vim.api.VimEditor
import com.maddyhome.idea.vim.api.injector
import com.maddyhome.idea.vim.command.Command
import com.maddyhome.idea.vim.command.OperatorArguments
import com.maddyhome.idea.vim.common.TextRange
import com.maddyhome.idea.vim.group.visual.VimSelection
import com.maddyhome.idea.vim.handler.VisualOperatorActionHandler

/**
 * @author vlan
 */
@CommandOrMotion(keys = ["y"], modes = [Mode.VISUAL])
class YankVisualAction : VisualOperatorActionHandler.SingleExecution() {
  override val type: Command.Type = Command.Type.COPY

  override fun executeForAllCarets(
    editor: VimEditor,
    context: ExecutionContext,
    cmd: Command,
    caretsAndSelections: Map<VimCaret, VimSelection>,
    operatorArguments: OperatorArguments,
  ): Boolean {
    val selections = caretsAndSelections.values
    val starts: MutableList<Int> = ArrayList()
    val ends: MutableList<Int> = ArrayList()
    selections.forEach { selection: VimSelection ->
      val textRange = selection.toVimTextRange(false)
      textRange.startOffsets.forEach { e: Int -> starts.add(e) }
      textRange.endOffsets.forEach { e: Int -> ends.add(e) }
    }
    val vimSelection = selections.firstOrNull() ?: return false
    val startsArray = starts.toIntArray()
    val endsArray = ends.toIntArray()
    return injector.yank.yankRange(editor, context, TextRange(startsArray, endsArray), vimSelection.type, true)
  }
}
